src/fs/bundle.rs

```diff
@@ -0,0 +1,235 @@
+    }
+
+    pub fn open_host_dir(path: &Path) -> Result<BundleData, std::io::Error> {
+        Ok(BundleData::HostDirectory(path.to_path_buf()))
+    }
+
+    pub fn open_ipa(path: &Path) -> Result<BundleData, ZipError> {
+        let file = std::fs::File::open(path)?;
+        let mut zip = ZipArchive::new(file)?;
+        let bundle_path = Self::find_bundle_path_in_archive(&mut zip)?;
+        Ok(BundleData::Zip { zip, bundle_path })
+    }
+
+    pub fn open_any(path: &Path) -> Result<BundleData, std::io::Error> {
+        if path.is_file()
+            && path
+                .extension()
+                .map(|ext| ext.eq_ignore_ascii_case("ipa"))
+                .unwrap_or(false)
+        {
+            Ok(Self::open_ipa(path)?)
+        } else if path.is_dir() {
+            Ok(Self::open_host_dir(path)?)
+        } else {
+            Err(std::io::Error::new(
+                std::io::ErrorKind::InvalidInput,
+                "not a directory or .ipa file",
+            ))
+        }
+    }
+
+    pub(super) fn into_fs_node(self) -> FsNode {
+        match self {
+            BundleData::HostDirectory(path) => FsNode::from_host_dir(&path, false),
+            BundleData::Zip { zip, bundle_path } => {
+                let archive = Arc::new(Mutex::new(zip));
+
+                let mut archive_guard = archive.lock().unwrap();
+
+                let mut builder = FsNodeBuilder::new();
+                for i in 0..archive_guard.len() {
+                    let file = archive_guard.by_index(i).unwrap(); // TODO: report IO error?
+                    let name = file.name();
+                    if let Some(path) = name.strip_prefix(&bundle_path) {
+                        let path = GuestPath::new(path);
+                        if file.is_dir() {
+                            builder.add_directory(&path);
+                        } else {
+                            builder.add_file(
+                                &path,
+                                FsNode::bundle_zip_file(IpaFileRef {
+                                    archive: archive.clone(),
+                                    index: i,
+                                }),
+                            );
+                        }
+                    }
+                }
+                builder.build()
+            }
+        }
+    }
+
+    pub fn read_plist(&mut self) -> Result<Vec<u8>, std::io::Error> {
+        match self {
+            BundleData::HostDirectory(path) => std::fs::read(path.join("Info.plist")),
+            BundleData::Zip { zip, bundle_path } => {
+                let mut file = zip.by_name(&format!("{}/Info.plist", bundle_path))?;
+                let mut buf = Vec::new();
+                file.read_to_end(&mut buf)?;
+                Ok(buf)
+            }
+        }
+    }
+}
+
+/// Represents a file inside an IPA bundle that can be opened.
+#[derive(Debug)]
+pub struct IpaFileRef {
+    archive: Arc<Mutex<ZipArchive<std::fs::File>>>,
```

Comment:
Currently this project is _very_ single-threaded, even though it simulates a potentially multi-threaded app, so it'd be fine to use `Rc` instead. I assume we can easily change it back in future? I'm not sure about when it's necessary to use a `Mutex` in Rust, but I assume it has a single-threaded replacement too.

Reply:
Hmmm, how are gonna implement threading btw? I tried to design a threading system in my HLE and found it... Non-trivial, so to say

Reply:
There's already a very primitive threading system since Super Monkey Ball needs it. I can explain a bit about how it works in some other channel (I don't want to get too off-topic here), but you can see the guts of it in `main.rs`. Incidentally, I really should move it out of that file, it's getting too complicated.